pkg/controller/keyvault/secret/keyvault_secret.go

```diff
@@ -0,0 +1,192 @@
+	"github.com/crossplane/crossplane-runtime/pkg/ratelimiter"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+
+	"github.com/crossplane/provider-azure/apis/keyvault/v1alpha1"
+	azure "github.com/crossplane/provider-azure/pkg/clients"
+	secretclients "github.com/crossplane/provider-azure/pkg/clients/keyvault/secret"
+)
+
+const (
+	errNotSecret            = "the custom resource is not a Secret instance"
+	errUpdateSecretCRFailed = "cannot update Secret custom resource instance"
+	errCheckUpToDate        = "unable to determine if external resource is up to date"
+
+	errConnectFailed = "cannot connect to Azure API"
+	errGetFailed     = "cannot get Key Vault Secret from Azure API"
+	errCreateFailed  = "cannot create the Key Vault Secret"
+	errUpdateFailed  = "cannot update the Key Vault Secret"
+	errDeleteFailed  = "cannot delete the Key Vault Secret"
+)
+
+// SetupSecret adds a controller that reconciles KeyVaultSecret resources.
+func SetupSecret(mgr ctrl.Manager, l logging.Logger, rl workqueue.RateLimiter, poll time.Duration) error {
+	name := managed.ControllerName(v1alpha1.KeyVaultSecretGroupKind)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		WithOptions(controller.Options{
+			RateLimiter: ratelimiter.NewDefaultManagedRateLimiter(rl),
+		}).
+		For(&v1alpha1.KeyVaultSecret{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(v1alpha1.KeyVaultSecretGroupVersionKind),
+			managed.WithExternalConnecter(&connector{kube: mgr.GetClient()}),
+			managed.WithPollInterval(poll),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+type connector struct {
+	kube client.Client
+}
+
+func (c connector) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	_, auth, err := azure.GetAuthInfo(ctx, c.kube, mg)
+	if err != nil {
+		return nil, errors.Wrap(err, errConnectFailed)
+	}
+	cl := keyvault.New()
+	cl.Authorizer = auth
+	return &external{kube: c.kube, client: cl}, nil
+}
+
+type external struct {
+	kube   client.Client
+	client keyvaultapi.BaseClientAPI
+}
+
+func (c *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1alpha1.KeyVaultSecret)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotSecret)
+	}
+
+	secret, err := c.client.GetSecret(ctx, cr.Spec.ForProvider.VaultBaseURL, cr.Spec.ForProvider.Name, "" /* latest */)
+	if err != nil {
+		return managed.ExternalObservation{ResourceExists: false}, errors.Wrap(resource.Ignore(azure.IsNotFound, err), errGetFailed)
+	}
+
+	lateInit := false
+	currentSpec := cr.Spec.ForProvider.DeepCopy()
+	secretclients.LateInitialize(&cr.Spec.ForProvider, secret)
+	if !cmp.Equal(currentSpec, &cr.Spec.ForProvider) {
+		if err := c.kube.Update(ctx, cr); err != nil {
```

Comment:
No need to update the MR (with late-initialized fields) here. Because the managed reconciler will [take care of it](https://github.com/crossplane/crossplane-runtime/blob/9f3f799fbd869778d892b506c7f2e789bbbf2f43/pkg/reconciler/managed/reconciler.go#L879) as long as we set `managed.ExternalObservation.ResourceLateInitialized` and `managed.ExternalObservation.ResourceExists` appropriately.